test/rules/join_return_with_assignment.dart

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N join_return_with_assignment`
+
+int bar() {
+  int a = 3;
+  --a; // LINT
+  return a;
+}
+
+int foo() {
+  int a = 3;
+  a++; // LINT
+  return a;
+}
+
+int doubleFoo() {
+  int a = 3;
+  a += 5;
+  a += 15; // OK
+  return a;
+}
```

Comment:
It is because there is some code, specially when calculating the hashcode that is like this:

```dart
int get hashCode {
  int hash = object.hashCode;
  hash <<= 2;
  hash &=  object2.hashCode;
  hash *= 32;
  ...
  hash ^= 15; // OK
  return hash;
}
```

So it would be highly probable that the user would like to add a new assignment between the return and the last assignation statement, so if you lint it and you solve the lint, then adding a new one would take a higher cost for the user, that was what I saw running the rule in the sdk